### DISCUSSION TITLE
"Plugin telescope-nvim not found", even when I can do :packadd telescope-nvim

### DISCUSSION THREAD

**Discussion by viperML** [nvim-neorocks/lz.n#45] [Q&A] [answered]
Hello!

I'm currently rewriting all of nixpkgs' wrappers around neovim, so I just started to try lz.n...

The issue is simple: after configuring lz with the following:

```lua
  {
    "telescope-nvim",
    cmd = "Telescope",
    after = function()
      vim.notify("Telescope loaded")
    end,
  },
```

And running `:Telescope`, I will get an error saying `Plugin telescope-nvim` not found, even when I can do `:packadd telescope-nvim`.

Interestingly, removing the `cmd` will work fine!

You can run my neovim wrapper with: `nix build github:viperML/dotfiles?ref=lz#neovim`, the relevant files are here:

- Wrapper: (you might want to inspect the resulting nvim binary if needed) https://github.com/viperML/dotfiles/blob/lz/modules/wrapper-manager/neovim/default.nix
- Init: https://github.com/viperML/dotfiles/blob/lz/modules/wrapper-manager/neovim/viper-init-plugin/lua/viper/init.lua#L18

Thank you for your time

**Comment by mrcjkb** [comment]
Hey :wave: 

The plugin's name is `"telescope.nvim"`, not `telescope-nvim` (in nixpkgs, the `pname` is `"telescope.nvim"`, which is the plugin's name that should be passed to `packadd`).

I have no idea why `packadd telescope-nvim` is working for you.

**Comment by viperML** [comment]
I'm not using nixpkgs' vimPlugins, but rather packaging them by myself, and I'm using `telescope-nvim` as the name of the directory in opt:

```
$ nix build "github:viperML/dotfiles?ref=lz#neovim.packDir"

$ file ./result/pack/viper-pack/opt/telescope-nvim
./result/pack/viper-pack/opt/telescope-nvim: symbolic link to /nix/store/hl0nv3rg24vy88fn2hawc99s14w0xd7b-telescope-nvim-0.1.8
```

I don't know know if there can be problems by not matching the dir name to upstream's repository, but as I said, `:packadd telescope-nvim` works for this setup.

**Comment by mrcjkb** [comment]
The `Plugin <plugin> not found` error indicates that the plugin may have been loaded, but not properly unregistered from the `cmd` handler.

I noticed you have `require("lzn-auto-require.loader").register_loader()` in your init.lua. Does it work if you comment that line?

**Comment by viperML** [comment]
I'm afraid it's the same...

**Reply by mrcjkb** [reply]
What's the output of `:lua =require("lz.n.state").plugins` before executing the `:Telescope` command?

**Reply by viperML** [reply]
The empty table `{}`

**Reply by viperML** [reply]
If you want to try reproduce locally, `nix build "github:viperML/dotfiles?ref=lz#neovim"` should give you the same setup, and I've tried hard for it to ignore any configuration at ~/.config

**Reply by mrcjkb** [reply]
> If you want to try reproduce locally

Could you please provide a minimal repro?

**Reply by viperML** [reply]
Everything seems to be ok with a minimal reproducer, so I will keep investigating what happened at my repo:

```nix
let
  pkgs = import (fetchTarball "https://github.com/NixOS/nixpkgs/archive/00d80d13810dbfea8ab4ed1009b09100cca86ba8.tar.gz") {};
  neovim = pkgs.neovim-unwrapped;

  lz-n = pkgs.fetchFromGitHub {
    owner = "nvim-neorocks";
    repo = "lz.n";
    rev = "v1.4.4";
    sha256 = "sha256-EAs/0whx8lsM1IRbdJ87WQs/N92Yovz+J15YKJf3OCo=";
  };

  plenary-nvim = pkgs.fetchFromGitHub {
    owner = "nvim-lua";
    repo = "plenary.nvim";
    rev = "a3e3bc82a3f95c5ed0d7201546d5d2c19b20d683";
    sha256 = "sha256-5Jf2mWFVDofXBcXLbMa417mqlEPWLA+cQIZH/vNEV1g=";
  };

  telescope-nvim = pkgs.fetchFromGitHub {
    owner = "nvim-telescope";
    repo = "telescope.nvim";
    rev = "0.1.8";
    sha256 = "sha256-e1ulhc4IIvUgpjKQrSqPY4WpXuez6wlxL6Min9U0o5Q=";
  };

  packDir = pkgs.runCommandLocal "pack-dir" {} ''
    mkdir -pv $out/pack/p/{start,opt}

    ln -vsT ${lz-n} $out/pack/p/start/lz-n
    ln -vsT ${telescope-nvim} $out/pack/p/opt/telescope-nvim
    ln -vsT ${plenary-nvim} $out/pack/p/opt/plenary-nvim

    mkdir -pv $out/pack/p/start/viper-init/plugin
    tee $out/pack/p/start/viper-init/plugin/init.lua <<EOF
      require("lz.n").load {
        {
          "plenary-nvim",
        },
        {
          "telescope-nvim",
          cmd = "Telescope",
        },
      }
    EOF
  '';
in
  pkgs.symlinkJoin {
    name = "neovim";
    paths = [neovim];
    nativeBuildInputs = [pkgs.makeWrapper];
    postBuild = ''
      wrapProgram $out/bin/nvim \
        --set NVIM_APPNAME "nvim-viper-2" \
        --add-flags "-u" \
        --add-flags "NORC" \
        --add-flags "--cmd" \
        --add-flags '"set packpath^=${packDir} | set runtimepath^=${packDir}"'
    '';
  }
```

**Reply by viperML** [reply]
Aha, I found the issue @mrcjkb 

```
      require("lz.n").load {
        {
          "telescope-nvim",
          cmd = "Telescope",
          after = function()
            vim.notify("telescope loaded")
          end
        },
      }
      require("lz.n").load {

      }
```

Calling load a second time is how you can reproduce it. If you comment out the `cmd`, then telescope loads fine and the notification triggers.

**Reply by mrcjkb** [reply]
🤔 where is `load` being called a second time in your config?
Calling `load { }` (with an empty list) is definitely undefined behaviour.
It would probably be better if lz.n threw an error in that case.

**Reply by viperML** [reply]
I was preparing to set-up a lazy plugin (premature optimization eh?), and these were the two calls https://github.com/viperML/dotfiles/blob/cf358a0b91d4cafcc9bcf7e0dd63ae7f8eca8f11/modules/wrapper-manager/neovim/default.nix#L102-L106 and https://github.com/viperML/dotfiles/blob/cf358a0b91d4cafcc9bcf7e0dd63ae7f8eca8f11/modules/wrapper-manager/neovim/viper-init-plugin/lua/viper/init.lua#L15-L32 . I guess I can merge them without any hassle.

> Calling load { } (with an empty list) is definitely undefined behaviour. It would probably be better if lz.n threw an error in that case.

So the problem is calling load a second time, or calling load on the empty list?

In any case, I agree that throwing an error and exiting is the better UX.

**Reply by mrcjkb** [reply]
> So the problem is calling load a second time, or calling load on the empty list?

I'd have to check, but I think it was intended to either be called once on a list of specs or multiple times on a single spec.
I didn't think someone would want to do both, but maybe I can make that possible.

**Reply by mrcjkb** [reply, marked as answer]
This should be fixed when #56, which removes the `lz.n.state` module as a redundant source of truth, is merged.

**Reply by viperML** [reply]
Thank you for your time
